rust-runtime/aws-smithy-observability/src/meter.rs

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Metrics are used to gain insight into the operational performance and health of a system in
+//! real time.
+
+use crate::attributes::{Attributes, Context};
+use crate::error::ObservabilityError;
+
+/// Provides named instances of [Meter].
+pub trait MeterProvider {
+    /// Get or create a named [Meter].
+    fn get_meter(&self, scope: &'static str, attributes: Option<&Attributes>) -> Box<dyn Meter>;
+
+    /// Optional method to flush the metrics pipeline, default is noop
+    fn flush(&self) -> Result<(), ObservabilityError> {
+        Ok(())
+    }
+
+    /// Optional method to shutdown the metrics provider, default is noop
+    fn shutdown(&self) -> Result<(), ObservabilityError> {
+        Ok(())
+    }
+}
+
+/// The entry point to creating instruments. A grouping of related metrics.
+pub trait Meter {
+    /// Create a new Gauge.
+    #[allow(clippy::type_complexity)]
+    fn create_gauge(
+        &self,
+        name: String,
```

Comment:
These should probably be more flexible like `impl Into<String>` or `impl Into<Cow<'static, str>>`, etc.

Reply:
I tried that, but once again it makes the `Meter` trait no longer object safe